src/java/org/apache/cassandra/db/guardrails/Guardrails.java

```diff
@@ -0,0 +1,175 @@
+                      format("Tables cannot have more than %s secondary indexes, failed to create secondary index %s",
+                             threshold, what));
+
+    /**
+     * Guardrail on the number of materialized views per table.
+     */
+    public static final Threshold materializedViewsPerTable =
+    factory.threshold("materialized_views_per_table",
+                      () -> -1,
+                      config::getMaterializedViewsPerTableFailThreshold,
+                      (isWarning, what, value, threshold) ->
+                      format("Tables cannot have more than %s materialized views, failed to create materialized view %s",
+                             threshold, what));
+
+    /**
+     * Guardrail ignoring/disallowing the usage of certain table properties.
+     */
+    public static final Values<String> allowedTableProperties =
+    factory.values("allowed_table_properties",
+                   config::getTablePropertiesIgnoredAsSet,
+                   config::getTablePropertiesDisallowedAsSet,
+                   String::toLowerCase,
+                   "Table Properties");
+
+    /**
+     * Guardrail disabling user-provided timestamps.
+     */
+    public static final DisableFlag userTimestampsEnabled =
+    factory.disableFlag("user_timestamps_enabled",
+                        () -> !config.getUserTimestampsEnabled(),
+                        "User provided timestamps (USING TIMESTAMP)");
+
+    private Guardrails()
+    {
+    }
+
+    /**
+     * Whether guardrails are enabled.
+     *
+     * @return {@code true} if guardrails are enabled and daemon is initialized,
+     * {@code false} otherwise (in which case no guardrail will trigger).
+     */
+    public static boolean enabled()
+    {
+        return config.getEnabled() && DatabaseDescriptor.isDaemonInitialized();
+    }
+
+    /**
+     * Register a {@link Listener}.
+     *
+     * <p>Note that listeners are called in the order they are registered, and on the thread on which the guardrail
+     * is triggered.
+     *
+     * @param listener the listener to register. If the same listener is registered twice (or more), its method will be
+     *                 called twice (or more) for every trigger.
+     */
+    public static void register(Listener listener)
+    {
+        listeners.add(listener);
+    }
+
+    /**
+     * Unregister a previously registered listener.
+     *
+     * @param listener the listener to unregister. If it was not registered before, this is a no-op. If it was
+     *                 registered more than once, only one of the instances is unregistered.
+     */
+    public static void unregister(Listener listener)
+    {
+        listeners.remove(listener);
+    }
+
+    /**
+     * Interface for external listeners interested in being notified when a guardrail is triggered.
+     *
+     * <p>Listeners should be registered through the {@link #register} method to take effect.
+     */
+    public interface Listener
```

Comment:
as already mentioned, if we absolutely want to go with listeners, I would provide default listener which would integrate with diagnostics framework and it would emit some events. I dont see any default implementation of listener yet. That might be the default one. Indeed, diagnostics framework itself would need to be enabled in order to actually see some events created.